src/utils/logger.ts

```diff
@@ -0,0 +1,68 @@
+import { AddressOrPair } from "@polkadot/api/submittable/types";
+import { SubmittableExtrinsic, ApiTypes } from "@polkadot/api/types";
+import { Callback, ISubmittableResult } from "@polkadot/types/types";
+import { EventRecord, DispatchError } from "@polkadot/types/interfaces/system";
+import { ApiPromise } from "@polkadot/api";
+
+declare type AnyFunctionReturningSubmittableExtrinsicApis =
+    SubmittableExtrinsic<ApiTypes>;
+
+interface EventContainer {
+    events: EventRecord[];
+}
+
+export async function sendLoggedTx(
+    transaction: AnyFunctionReturningSubmittableExtrinsicApis,
+    signer: AddressOrPair,
+    api: ApiPromise,
+    eventContainer: EventContainer
+): Promise<void> {
+    // When passing { nonce: -1 } to signAndSend the API will use system.accountNextIndex to determine the nonce
+    const unsubscribe: Callback<ISubmittableResult> = await transaction
+        .signAndSend(
+            signer, { nonce: -1 }, (result) => txCallback(unsubscribe, result, api, eventContainer)
+        ) as Callback<ISubmittableResult>;
+}
+
+function txCallback(
+    unsubscribe: Callback<ISubmittableResult>,
+    result: ISubmittableResult,
+    api: ApiPromise,
+    eventContainer: EventContainer
+) {
+    if (result.status.isFinalized) {
+        console.log(`Transaction finalized at blockHash ${result.status.asFinalized}`);
+        unsubscribe(result);
+        printEvents(result.events, api);
+        eventContainer.events = result.events;
+    }
+}
+
+function printEvents(events: EventRecord[], api: ApiPromise) {
+    let foundErrorEvent = false;
+    events.forEach(({ event }) => {
```

Comment:
Nitpick: if you want to avoid a bit of nesting, you can probably go with

```typescript
events.flatMap(({ event }) => event.data).forEach( (eventData: any) =>)
```

it will require to add
```json
"lib": [
    "ES2019"
]
```

to the tsconfig.json `compilerOptions` part, which I think is fine.

Same comment as above for early return.